### PR TITLE
Work on higher-order MappingFE

### PR DESCRIPTION
**Pull request by peterrum**

### CODE REVIEW

File: source/fe/mapping_fe.cc [hunk 1]
```diff
@@ -2142,17 +2141,21 @@ MappingFE<dim, spacedim>::compute_mapping_support_points(
...
+    fe->get_unit_support_points().size());
+
+  Table<2, double> weights;
+
+  cell->get_manifold().get_new_points(vertices,
+                                      weights,
+                                      mapping_support_points);
 
-  return a;
+  return mapping_support_points;
```
Comment (peterrum) [inline source/fe/mapping_fe.cc:2158]: @luca-heltai I think this matches what you described. But what I am not sure about how to integrate line and quad manifolds here, like it is done here:

https://github.com/dealii/dealii/blob/9f9205129e30e174594e3103136c5d8f39a2c635/source/fe/mapping_q_generic.cc#L1544-L1639

Comment (luca-heltai) [reply source/fe/mapping_fe.cc:2158]: For the moment, put an assertion that is thrown if the manifold ids of the subobjects are different than the one of the cell. What you wrote works for objects where everybody has the same manifold id. 

The idea would be to loop over all subobjects, querying (if any) all the support points that live on the edges, quads/tri, hex/tet/etc first, and then using the same thing you did here with the vertices of the subobjects and the support points that live in the middle of the subobjects.

Comment (peterrum) [reply source/fe/mapping_fe.cc]: I have added the code. What would be an appropriate manifold to test with?

Comment (luca-heltai) [reply source/fe/mapping_fe.cc]: Hyper shell, or cylinder_shell. They would work with a polar manifold, or a cylindrical manifold.

**Comment by peterrum** [comment]
Linear and quadratic mapping:

mapping-fe

FYI @luca-heltai @agrayver

**Comment by luca-heltai** [comment]
So, the high order output is now implemented also for tria/tets? Nice!

:)

File: source/fe/mapping_fe.cc [hunk 2]
```diff
@@ -2204,23 +2232,74 @@ MappingFE<dim, spacedim>::transform(
...
 
+namespace
+{
+  template <int spacedim>
+  bool
+  check_all_manifold_ids_identical(
+    const TriaIterator<CellAccessor<1, spacedim>> &)
+  {
+    return true;
+  }
```
Comment (luca-heltai) [inline source/fe/mapping_fe.cc:2243]: Please add two blank spaces, both here and below.

**Review by luca-heltai — APPROVED** [review]
Minor changes.

**Comment by peterrum** [comment]
> So, the high order output is now implemented also for tria/tets? Nice!

Indeed x2!

**Review by luca-heltai — APPROVED** [review]

File: source/grid/reference_cell.cc [hunk 3]
```diff
@@ -101,6 +104,42 @@ namespace ReferenceCell
       }
   }
 
+  template <int dim>
```
Comment (kronbichler) [inline source/grid/reference_cell.cc:107]: Please add two empty lines before this to have 3 as usual.

File: source/grid/reference_cell.cc [hunk 4]
```diff
@@ -101,6 +104,42 @@ namespace ReferenceCell
...
+  const ScalarPolynomialsBase<dim> &
+  get_linear_poly_space(const Type &reference_cell)
+  {
+    AssertDimension(dim, get_dimension(reference_cell));
+
+    static const TensorProductPolynomials<dim> poly(
+      Polynomials::generate_complete_Lagrange_basis(
+        QGaussLobatto<1>(2).get_points()));
```
Comment (kronbichler) [inline source/grid/reference_cell.cc:115]: You can have it slightly shorter and with fewer indirections in the implementation by
```suggestion
      Polynomials::LagrangeEquidistant::generate_complete_basis(1));
```

**Review by kronbichler — COMMENTED** [review]
A few additional small comments.

File: include/deal.II/grid/reference_cell.h [hunk 5]
```diff
@@ -507,6 +509,14 @@ namespace ReferenceCell
   make_triangulation(const Type &                  reference_cell,
                      Triangulation<dim, spacedim> &tria);
 
+
+  /**
+   * Return a linear nodal space matching the given reference cell.
+   */
+  template <int dim>
+  const ScalarPolynomialsBase<dim> &
+  get_linear_poly_space(const Type &reference_cell);
```
[the remark this answers is not in the document]
Comment (peterrum) [reply include/deal.II/grid/reference_cell.h]: You are right. I have removed the new function and extended the existing one.

**Review by kronbichler — APPROVED** [review]

**Pull request merged**